### PR TITLE
Implement snap interaction

### PR DESCRIPTION
**Pull request by lehecht**
Implements snap interaction to enable creating shapes in different 45° angles.

To activate snapping during drawing, the control key needs to be pressed.

### CODE REVIEW

**Comment by mzur** [comment]
Resolves #252

**Review by mzur — COMMENTED** [review]
A creative approach with the visible "guidelines"! But again, I probably should have made more clear how I envision this feature. I'll demonstrate how I think it should look with Inkscape (drawing a polygon but this also transfers to rectangles):

[Screencast from 05.03.2024 09:39:34.webm](https://github.com/biigle/core/assets/2457311/2e38f138-4cfb-4638-a02c-f59d77b75a34)

In Inkscape, while I press <kbd>Ctrl</kbd>, the line snaps to multiples of 15° (here we still want only 45°). When I click to create a new point of the line, the next line segment shows the same behavior.

The difference to the current approach is:

- There are no visible guidelines.
- I need to keep the shortcut key pressed to activate the behavior. When the key is released, the current line segment no longer snaps to the angles.
- It works individually for each line segment. Currently, the guidelines are only created for the first.

While you probably focused on rectangles with this first implementation, the behavior described above would (and should) also work on lines and polygons because the interactions are the same.

I think the current implementation is not too far off. Using guidelines and the snap interaction is the way to go. The guidelines could be made invisible and have to be updated for each new point. Then the whole thing should only be active while <kbd>Ctrl</kbd> is pressed.

**Comment by lehecht** [comment]
@mzur 
I noticed that the snapping not only occurs for the guidlines but also for other annotations. Is this a problem that needs to be fixed?

**Comment by mzur** [comment]
> I noticed that the snapping not only occurs for the guidlines but also for other annotations. Is this a problem that needs to be fixed?

Yes. You can use the [`source` option](https://openlayers.org/en/v5.3.0/apidoc/module-ol_interaction_Snap-Snap.html) of the snap interaction to control this.

**Comment by lehecht** [comment]
@mzur 
Is there a reason why you declared [this function](https://github.com/biigle/core/blob/662e38e4c92e9bb2c72ce133dfb628451c804b88/resources/assets/js/annotations/components/annotationCanvas/drawInteractions.vue#L17-L25) not as vue-method? I'm currently fixing a bug and the solution could be either to make this function a vue-method, in order to change a data property there, or to use a separate shift-key-event listener in the created method.

**Comment by mzur** [comment]
> Is there a reason why you declared [this function](https://github.com/biigle/core/blob/662e38e4c92e9bb2c72ce133dfb628451c804b88/resources/assets/js/annotations/components/annotationCanvas/drawInteractions.vue#L17-L25) not as vue-method?

Only that there was no need for it before. You can implement it as a Vue method, too.

**Comment by lehecht** [comment]
Note: My last bug fix only works when biigle/openlayers#5 is merged.

**Review by mzur — CHANGES_REQUESTED** [review]
We are getting there!

It seems like I have to press <kbd>Ctrl</kbd> before drawing the first point of the annotation to enable this feature. Can you change this so I can press the key anytime (even after the first point) and the feature is enabled?

Can you also increase the `pixelTolerance` of the snap interaction to maybe 15 or so?

File: resources/assets/js/annotations/snapInteraction.vue [hunk 1]
```diff
@@ -85,9 +91,10 @@ export default {
             return res;
         },
         startSnap() {
-            this.snapInteraction = new Snap({ source: this.annotationSource });
-            this.map.addInteraction(this.snapInteraction);
             this.drawSnaplines();
+            this.snapInteraction = new Snap({ source: this.annotationSource, features: new Collection(this.snapLineFeatures) });
```
[the remark this answers is not in the document]
Comment (lehecht) [reply resources/assets/js/annotations/snapInteraction.vue:95]: When I delete the annotationSource and feature styling, the snap lines are still visible. Did you just remove the annotation Source from line 95 or did you do also something else?

**Comment by lehecht** [comment]
> It seems like I have to press Ctrl before drawing the first point of the annotation to enable this feature. Can you change this so I can press the key anytime (even after the first point) and the feature is enabled?

I don't exactly know what you mean. When I create the first point and press Ctrl afterwards, I can use the tool without problems. But I need to say that the old and smaller pixelTolerance made it a bit more difficult to distinguish if that tool was active or not.

**Comment by mzur** [comment]
> When I create the first point and press Ctrl afterwards, I can use the tool without problems.

You are right. Maybe I didn't notice the snapping for some reason...

Can you please also add something to the keyboard shortcuts manual articles (image and video)? Then I'll do my final review but everything looks good so far!

**Review by mzur — COMMENTED** [review]
The feature is perfect :ok_hand: I've just commented on some minor things regarding the code. Also, I've updated the manual articles a little, as I found that the shortcut should be moved to a separate "section" there.

**Comment by mzur** [comment]
Here is also a follow-up issue: https://github.com/biigle/core/issues/836

File: resources/assets/js/annotations/components/annotationCanvas/drawInteractions.vue [hunk 2]
```diff
@@ -88,11 +97,30 @@ export default {
...
+                drawInteraction.on('drawabort', () => {
+                        this.drawEnded = true;
+                    });
             }
         },
+        updateSnapCoords(mapBrowserEvent) {
+            this.snappingCoords = mapBrowserEvent.coordinate;
+            this.shouldSnap = mapBrowserEvent.originalEvent.ctrlKey;
+            return true;
+        }
```
[the remark this answers is not in the document]
Comment (lehecht) [reply resources/assets/js/annotations/components/annotationCanvas/drawInteractions.vue:122]: My current approach looks like this:
```
window.addEventListener('click', (e) => {
            if(this.drawInteraction?.getActive() && this.drawInteraction?.sketchCoords_?.length > 0){
                this.shouldSnap = e.ctrlKey;
                let idx = this.drawInteraction.sketchCoords_.length - 2;
                this.snappingCoords = this.drawInteraction.sketchCoords_[idx];
            }
        });
```

**Review by mzur — APPROVED** [review]
Great, thanks!

**Pull request merged**